### PR TITLE
K8SPS-105 fix sidecar error output

### PR DESCRIPTION
**Pull request by pooknull**
[K8SPS-87](https://jira.percona.com/browse/K8SPS-87) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

Fixed error output in `backupHandler`. The problem with the sidecar was that we were trying to read from `xbcloudErr` in a new goroutine, but if we retrun in the main one before `<-xbcloudDone`, it's not possible because of `defer xbcloudErr.Close()`. And the second problem is that we could write the status code 2 times to `ResponseWriter` in separate goroutines what caused `http: superfluous response.WriteHeader call` error message

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/auto-config.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/demand-backup.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/init-deploy.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/gr-init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/semi-sync.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/scaling.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/config.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/service-per-pod.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/users.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/sidecars.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/3f5317e/limits.log) |
| We run 12 out of 12|

commit: https://github.com/percona/percona-server-mysql-operator/pull/157/commits/3f5317e232439086c63c9ada5b5cfe2c7d441414
image: `perconalab/percona-server-mysql-operator:PR-157-3f5317e`

File: controllers/perconaservermysqlbackup_controller.go [hunk 1]
```diff
@@ -298,20 +300,88 @@ func getDestination(storage *apiv1alpha1.BackupStorageSpec, clusterName, creatio
...
+	status, source, err := db.ReplicationStatus()
+	if err != nil {
+		return "", errors.Wrap(err, "check replication status")
+	}
+	primary := ""
+	if status == replicator.ReplicationStatusActive {
+		primary = source
+	}
+	return primary, nil
+}
```
Comment (egegunes) [inline controllers/perconaservermysqlbackup_controller.go:387]: can we move these functions to a separate package and use them in `cmd/bootstrap` too?

Comment (egegunes) [reply controllers/perconaservermysqlbackup_controller.go:387]: `pkg/mysql/topology` could be the package

File: controllers/perconaservermysqlbackup_controller.go [hunk 2]
```diff
@@ -298,20 +300,88 @@ func getDestination(storage *apiv1alpha1.BackupStorageSpec, clusterName, creatio
...
+		for i := 0; i < int(cluster.Spec.MySQL.Size); i++ {
+			podIPs = append(podIPs, mysql.FQDN(cluster, i))
+		}
+		primary, replicas, err = getTopology(podIPs, operatorPass)
+		if err != nil {
+			return "", errors.Wrap(err, "select donor")
+		}
+	}
+	src := ""
+	l.Info("get topology", "primary", primary, "replicas", replicas)
```
Comment (egegunes) [inline controllers/perconaservermysqlbackup_controller.go:334]: it'd be better to make this log `DEBUG`
```suggestion
	l.V(1).Info("get topology", "primary", primary, "replicas", replicas)
```

**Review by egegunes — CHANGES_REQUESTED** [review]

File: controllers/perconaservermysqlbackup_controller.go (hunk 1 above)
Comment (pooknull) [reply controllers/perconaservermysqlbackup_controller.go]: https://github.com/percona/percona-server-mysql-operator/pull/157/commits/4c0b10032a5b2cb1d8835be51c25d6d5a8c2d51b

File: controllers/perconaservermysqlbackup_controller.go (hunk 2 above)
Comment (pooknull) [reply controllers/perconaservermysqlbackup_controller.go]: https://github.com/percona/percona-server-mysql-operator/pull/157/commits/4c0b10032a5b2cb1d8835be51c25d6d5a8c2d51b

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/demand-backup.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/init-deploy.log) |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/auto-config.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/gr-init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/semi-sync.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/scaling.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/config.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/service-per-pod.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/users.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/sidecars.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/4c0b100/limits.log) |
| We run 12 out of 12|

commit: https://github.com/percona/percona-server-mysql-operator/pull/157/commits/4c0b10032a5b2cb1d8835be51c25d6d5a8c2d51b
image: `perconalab/percona-server-mysql-operator:PR-157-4c0b100`

**Review by inelpandzic — COMMENTED** [review]
Hey @egegunes  , just curious: the orchestrator is highly available, and it can go completely down nevertheless, but are the chances of it actually going down high enough that we have to do this insurance? 

Also, we depend on the orchestrator for other things, what about those if the thing goes down?

File: pkg/mysql/topology.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	} else if primary == "" {
+		primary = replicas.List()[0]
+	}
+	if replicas.Len() > 0 {
+		replicas.Delete(primary)
+	}
+	return primary, replicas.List(), nil
+}
+
+func getPrimary(host, operatorPass string) (string, error) {
```
Comment (hors) [inline pkg/mysql/topology.go:36]: @pooknull Are you sure that  this check will work in case of external replication to our PRIMARY ?

**Review by hors — CHANGES_REQUESTED** [review]

File: pkg/mysql/topology.go (hunk 3 above)
Comment (inelpandzic) [reply pkg/mysql/topology.go:36]: Also, why not use `orchestrator.ClusterPrimary(..)` function?

**Comment by egegunes** [comment]
@inelpandzic @hors I propose reverting the changes related to orchestrator from this PR and merge only fixes for issues that @pooknull found.

We're basically asking @pooknull to use Orchestrator APIs in a PR with title "Don't depend on orchestrator for backups" and also asking why he don't use a function that the sole purpose of PR is to delete it: https://github.com/percona/percona-server-mysql-operator/pull/157/files#diff-0087813983dcee470abda3f96fdffd6beba8c8c42385bd1a2995a5c7e7251f84L302

Let's decide how should we depend on Orchestrator and do the development, this is going nowhere now.

> Hey @egegunes , just curious: the orchestrator is highly available, and it can go completely down nevertheless, but are the chances of it actually going down high enough that we have to do this insurance? 
>
> Also, we depend on the orchestrator for other things, what about those if the thing goes down?

@inelpandzic yes since we're running it in HA mode, the probability of Orchestrator going down is lower but it'll go down eventually. I don't think we should prevent depending on Orchestrator for functionality like backups (this ticket was not my idea, see [this comment](https://github.com/percona/percona-server-mysql-operator/pull/77#discussion_r894529722)). My view on Orchestrator is we shouldn't depend on it in things that decide MySQL pods faith: healthchecks, proxies, sidecars.

**Comment by hors** [comment]
> @inelpandzic @hors I propose reverting the changes related to orchestrator from this PR and merge only fixes for issues that @pooknull found.

@egegunes Can we wait for the proxy and get host for backup using it?

**Comment by inelpandzic** [comment]
> We're basically asking @pooknull to use Orchestrator APIs in a PR with title "Don't depend on orchestrator for backups" 
Yeah, total nonsense from me to ask.

Regarding the other considerations about the orchestrator, well it is just a continuation of yesterday's discussion we had, the dependence on in. I guess we will conclude something about that today. And with that I agree with:
> @inelpandzic @hors I propose reverting the changes related to orchestrator from this PR and merge only fixes for issues that @pooknull found.

**Comment by hors** [comment]
> @inelpandzic @hors I propose reverting the changes related to orchestrator from this PR and merge only fixes for issues that @pooknull found.
I support this idea. @pooknull please do it.
 
> We're basically asking @pooknull to use Orchestrator APIs in a PR with title "Don't depend on orchestrator for backups"
I did not ask to use Orchestrator APIs . I just asked the question about the replication ... And now we have the vision how it  should be. So, in the next release we will  add topology detection library to be used for backups, probes, proxy, bootstrap. We will have all logic in one place.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/demand-backup.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/init-deploy.log) |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/auto-config.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/gr-init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/semi-sync.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/scaling.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/service-per-pod.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/config.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/users.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/sidecars.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-157/e343b06/limits.log) |
| We run 12 out of 12|

commit: https://github.com/percona/percona-server-mysql-operator/pull/157/commits/e343b06a5eac022f9c53f1e310aa9a6ae7b2e0f1
image: `perconalab/percona-server-mysql-operator:PR-157-e343b06`

**Review by inelpandzic — APPROVED** [review]

**Pull request merged**